rerun_cpp/src/rerun/archetypes/segmentation_image_ext.cpp

```diff
@@ -1,75 +1,56 @@
+#include "../image_utils.hpp"
+
+    /// Row-major. Borrows.
     ///
-    /// \param shape
-    /// Shape of the image. Calls `Error::handle()` if the tensor is not 2-dimensional
-    /// Sets the dimension names to "height" and "width" if they are not specified.
-    /// \param buffer
-    /// The tensor buffer containing the segmentation image data.
-    SegmentationImage(Collection<datatypes::TensorDimension> shape, datatypes::TensorBuffer buffer)
-        : SegmentationImage(datatypes::TensorData(std::move(shape), std::move(buffer))) {}
+    /// The length of the data should be `W * H`.
+    template <typename TElement>
+    SegmentationImage(const TElement* pixels, components::Resolution2D resolution_)
+        : SegmentationImage{
+              reinterpret_cast<const uint8_t*>(pixels), resolution_, get_data_type(pixels)} {}
 
-    /// New segmentation image from tensor data.
+    /// Row-major.
     ///
-    /// \param data_
-    /// The tensor buffer containing the segmentation image data.
-    /// Sets the dimension names to "height" and "width" if they are not specified.
-    /// Calls `Error::handle()` if the tensor is not 2-dimensional
-    explicit SegmentationImage(components::TensorData data_);
+    /// The length of the data should be `W * H`.
+    template <typename TElement>
+    SegmentationImage(std::vector<TElement> pixels, components::Resolution2D resolution_)
+        : SegmentationImage{Collection<TElement>::take_ownership(std::move(pixels)), resolution_} {}
 
-    /// New segmentation image from dimensions and pointer to segmentation image data.
+    /// Row-major.
     ///
-    /// Type must be one of the types supported by `rerun::datatypes::TensorData`.
-    /// \param shape
-    /// Shape of the image. Calls `Error::handle()` if the tensor is not 2-dimensional
-    /// Sets the dimension names to "height", "width" and "channel" if they are not specified.
-    /// Determines the number of elements expected to be in `data`.
-    /// \param data_
-    /// Target of the pointer must outlive the archetype.
+    /// The length of the data should be `W * H`.
     template <typename TElement>
-    explicit SegmentationImage(Collection<datatypes::TensorDimension> shape, const TElement* data_)
-        : SegmentationImage(datatypes::TensorData(std::move(shape), data_)) {}
+    SegmentationImage(Collection<TElement> pixels, components::Resolution2D resolution_)
+        : SegmentationImage{pixels.to_uint8(), resolution_, get_data_type(pixels.data())} {}
+
+    /// Row-major. Borrows.
+    ///
+    /// The length of the data should be `W * H * data_type.size`
+    SegmentationImage(
+        const void* data_, components::Resolution2D resolution_,
+        components::ChannelDataType data_type_
+    )
+        : data{Collection<uint8_t>::borrow(data_, num_bytes(resolution_, data_type_))},
+          resolution{resolution_},
+          data_type{data_type_} {}
+
+    /// New depth image from an `ChannelDataType` and a pointer.
```

Comment:
that's not the variant that takes a pointer, it's the one that takes a (row major) collection!